### ISSUE TITLE
Webpack guidance does not explain how to launch as debug

### ISSUE THREAD

**Issue by itowlson** [microsoft/vscode#79956]
Issue Type: <b>Bug</b>

I am trying to webpack my extension as described at https://code.visualstudio.com/api/working-with-extensions/bundling-extension and the linked repositories, and I am having trouble testing my extension under the debugger.

I have tried two approaches, one based on the `webpack-extension` sample and one on the `vscode-references-view` referred to in the article.

If I set `launch.json > preLaunchTask` to `npm: watch`, then if I make an edit and hit F5, `webpack --watch` doesn't complete in time, and I find myself running the previous edit of the extension.

If I set `launch.json > preLaunchTask` to `npm: webpack`, then it takes many seconds (as my extension is quite big) for webpack to do its thing, so it becomes quite painful to test frequently.

It would be good to have clear and specific guidance on exactly how to configure `launch.json` and the NPM scripts to enable accurate but rapid testing in the bundled environment.

VS Code version: Code 1.37.1 (f06011ac164ae4dc8e753a3fe7f9549844d15e35, 2019-08-15T16:17:55.855Z)
OS version: Windows_NT x64 10.0.18912

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Xeon(R) CPU E5-1650 v2 @ 3.50GHz (12 x 3492)|
|GPU Status|2d_canvas: enabled<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>native_gpu_memory_buffers: disabled_software<br>oop_rasterization: disabled_off<br>protected_video_decode: unavailable_off<br>rasterization: enabled<br>skia_deferred_display_list: disabled_off<br>skia_renderer: disabled_off<br>surface_synchronization: enabled_on<br>video_decode: enabled<br>viz_display_compositor: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|63.93GB (38.61GB free)|
|Process Argv|--folder-uri file:///d%3A/GitHub/vscode-kubernetes-tools|
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (17)</summary>

Extension|Author (truncated)|Version
---|---|---
better-toml|bun|0.3.2
gitlens|eam|9.9.3
EditorConfig|Edi|0.13.0
vscode-pull-request-github|Git|0.10.0
terraform|mau|1.4.0
vscode-docker|ms-|0.7.0
duffle-vscode|ms-|0.3.0
python|ms-|2019.8.30787
remote-wsl|ms-|0.39.2
azure-account|ms-|0.8.4
csharp|ms-|1.21.1
Go|ms-|0.11.4
powershell|ms-|2019.5.0
vscode-typescript-tslint-plugin|ms-|1.2.2
quicktype|qui|12.0.46
java|red|0.47.0
vscode-yaml|red|0.5.1

</details>

**Comment by itowlson** [comment]
If it helps, the current state of webpack for the particular extension I'm asking about is at https://github.com/Azure/vscode-kubernetes-tools/pull/586.

**Comment by itowlson** [comment]
I also tried cloning https://github.com/microsoft/vscode-references-view to see how that works, because I noticed the webpack PR modified the background watch task, and got an error `Invalid problemMatcher reference: $ts-webpack`.  So that isn't what I'm missing either.

**Comment by itowlson** [comment]
I'm still stuck on this.  I have two extensions that are getting bug reports about high CPU load and `vsce` warnings about too many files, and I can't webpack them without fouling up my ability to test/debug them.  Please send advice!  Thanks!

**Comment by isidorn** [comment]
Hey sorry for not responding, this somehow slipped through my radar.
The preLaunchTask should not be the npm: webpack. That should be done before releasing, not before debugging the extension.
The correct one should be the watch If the watch doesn't complete in time than this is an issue with the tasks / debug and for that can you please file a new issue - that the preLaunchTask does not finish before debugging starts.

fyi @jrieken on advice since he wrote the weback tutorial

**Comment by jrieken** [comment]
`ts-webpack` is an extension that you need to install, also consider running web pack in watch mode to speed things up

**Comment by itowlson** [comment]
Hmm, using a `preLaunchTask` of `npm: watch` now doesn't seem to be working _at all_.  On a clean checkout it doesn't generate anything into `dist`, and after running `npm run compile` it doesn't reflect any updates.

My `package.json` looks like this:

```json
    "scripts": {
        "vscode:prepublish": "webpack --mode production",
        "compile": "webpack --mode none",
        "watch": "webpack --mode none --watch",
        "test-compile": "tsc -p ./",
        "postinstall": "node ./node_modules/vscode/bin/install",
        "test": "npm run compile && node ./node_modules/vscode/bin/test"
    },
    "devDependencies": {
        "@types/mocha": "^2.2.32",
        "@types/node": "^10.2.0",
        "@types/request": "^2.48.1",
        "file-loader": "^4.0.0",
        "gulp": "^4.0.2",
        "gulp-tslint": "^8.1.2",
        "mocha": "^4.1.0",
        "ts-loader": "^6.0.4",
        "ts-webpack": "1.0.0",
        "tslint": "^5.9.1",
        "typescript": "^3.5.2",
        "vscode": "^1.1.21",
        "webpack": "^4.35.2",
        "webpack-cli": "^3.3.5"
    },
```

My `.vscode/launch.json` looks like this:

```json
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Extension",
            "type": "extensionHost",
            "request": "launch",
            "runtimeExecutable": "${execPath}",
            "args": [
                "--extensionDevelopmentPath=${workspaceFolder}"
            ],
            "outFiles": [
                "${workspaceFolder}/dist/**/*.js"
            ],
            "preLaunchTask": "npm: watch",
            "skipFiles": [
                "<node_internals>/**/*.js"
            ]
        }
    ]
}
```

My `.vscode/tasks.json` looks like this:

```json
{
    "version": "2.0.0",
    "tasks": [
        {
            "type": "npm",
            "script": "watch",
            "problemMatcher": [
                "$ts-webpack"
            ],
            "isBackground": true,
            "presentation": {
                "echo": true,
                "reveal": "never",
                "panel": "shared",
                "showReuseMessage": true,
                "clear": false
            }
        }
    ]
}
```

I get a warning `Incorrect type. Expected "object".` on the `problemMatcher` line here but this is copied from the references sample (though it is not mentioned in the webpack guidance).

I can't see where I'm diverging from the references sample... but, as mentioned, the references sample doesn't work for me either.

I'm clearly doing something wrong, because this obviously works for lots of other people.  But I can't for the life of me see what.

@jrieken I thought the debug task would already run webpack in watch mode.  Are you saying that when I start a new VS Code session I should open a terminal and manually run `webpack --watch`?

**Comment by itowlson** [comment]
Oh, I see.  The `$ts-webpack` thing needs this: https://marketplace.visualstudio.com/items?itemName=eamodio.tsl-problem-matcher.  I'm afraid that wasn't obvious to me; perhaps it could be called out in the readme/documentation?

I assume everyone who contributes to my extension will need to to install this too?

**Comment by itowlson** [comment]
Okay, installing the `eamodio.tsl-problem-matcher` extension gets me back to the situation where hitting `F5` debugs the code as it was a couple of seconds before I hit `F5`.  If I save, count slowly to 3, and then hit `F5` it works - this is definitely progress...!

@jrieken Is there a way to force the extension to install when a contributor updates the project, e.g. as part of `npm install`?  Or at least to issue a meaningful warning/error on build, so that contributors know what they need to install?

**Comment by jrieken** [comment]
Yeah, there is `.vscode/extensions.json`, e.g this is how we use it in vscode: https://github.com/microsoft/vscode/blob/2724ca6c053585673ee9cb26527d18657a51dfa9/.vscode/extensions.json#L4

**Comment by itowlson** [comment]
@jrieken Belated thanks!  Love that you folks have anticipated this sort of issue.

@isidorn I don't know whether I'm being a doofus but I am definitely seeing stale code being debugged unless I explicitly save and wait a few seconds before hitting F5.  I will raise a separate issue for this.

**Comment by itowlson** [comment]
@isidorn Raised #82076 to track the watch issue (or to figure out where I am messing up!) - thanks for your help!
